### PR TITLE
KAFKA-5663; Fix LogDirFailureTest system test

### PR DESCRIPTION
**Pull request by lindong28**

### CODE REVIEW

**Comment by lindong28** [comment]
@apurvam @becketqin Can you help review it when you get time?

**Comment by lindong28** [comment]
Previously the system test failed because I forgot to provide logDirFailureChannel when instantiating the Log instance. Therefore the broker will observe NullPointerException when it attempts to insert a log directory name to the logDirFailureChannel. The patch fixes the problem by providing the logDirFailureChannel when constructing the log.

The patch also adds a test to verify the fix. The bug wasn't detected because previously the log directory failure was triggered when broker attempt to write to checkpoint file, in which case the logDirFailureChannel is not null. The added test will explicitly trigger log.roll() which will use log.logDirFailureChannel if there is IOException.

In addition, the patch will log error whenever there is new log directory failure event. Previously the log directory failure event will not be logged if KafkaStorageException is caught in e.g. `ReplicaManager.appendToLocalLog()`.

**Comment by becketqin** [comment]
@lindong28 Thanks for the patch. Can you run system test again and see if there are other failures?

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6420/
Test PASSed (JDK 7 and Scala 2.11).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6405/
Test PASSed (JDK 8 and Scala 2.12).

**Comment by ijuma** [comment]
I haven't done a full review, one comment: @RunWith breaks the tagging of integration tests due to a JUnit bug. So, better to avoid it if possible.

**Comment by apurvam** [comment]
Thanks for the patch @lindong28 . Would the previous system test have failed all the time? It seems so, given the nature of the error. If so, why wasn't it caught by the integration test?

**Comment by lindong28** [comment]
@becketqin Sure. I have run system test again with and without this patch. The tests can pass in both cases. See https://jenkins.confluent.io/job/system-test-kafka-branch-builder/1017/console and https://jenkins.confluent.io/job/system-test-kafka-branch-builder/1018/console.

@ijuma Thanks for the suggestion. I have updated the patch to avoid using Parameterized test.

@apurvam Good point. Here is the explanation.

The system test attempts to verify that 1) partition in the offline log directory will moved out of insync replica and controller will re-elect leader; and 2) partition in the good log directory will not be affected.

The bug doesn't cause problem for the first part of the test because broker will discover properly handle offline log directory when it attempts to write to checkpoint file. LogManager will write to `kafka-recovery-point-checkpoint` file 30 seconds after broker starts and every 60 seconds afterwards. I verified that this operation has correctly triggered `maybeAddLogFailureEvent` after NullPointerException is thrown.

The bug would occasionally cause problem for the second part of the test. The second part of the test will fail in the following scenario:

1) Selected broker A is initially a follower of `test_topic_2` partition `0`. And broker B is leader for both `test_topic_1` partition `0` and `test_topic_2` partition `0`.
2) The same replicaFetcherThread keeps fetching data for both partitions.
3) The test makes the log directory of the `test_topic_1` partition `0` offline on broker A. Then NullPointerException causes the ReplicaFetcherThread to die. Broker A will stop fetching data for both partitions and it will drop out of isr for `test_topic_2` partition `0`.
4) Then the test shutdowns the broker B and broker C. Broker A will become controller but `test_topic_2` partition `0` will be offline because the only replica on broker A is out of ISR.
5) Then the producer will fail to produce message to this partition and the test will fail.

Based on this explanation, the second part of the test will succeed as long as broker B is not fetching data for `test_topic_2` and `test_topic_2` from the same other broker, which is likely to be true.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6429/
Test PASSed (JDK 7 and Scala 2.11).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6414/
Test PASSed (JDK 8 and Scala 2.12).

File: core/src/main/scala/kafka/log/LogManager.scala [hunk 1]
```diff
@@ -568,6 +571,7 @@ class LogManager(logDirs: Array[File],
           log
         } catch {
           case e: IOException =>
+            error(s"Error while creating log for $topicPartition in dir ${dataDir.getAbsolutePath}", e)
```
Comment (ijuma) [inline core/src/main/scala/kafka/log/LogManager.scala]: You said that these exceptions would previously be logged only if they are caught. Can we not fix this in one place instead of multiple places? The former would be more robust.

File: core/src/test/scala/integration/kafka/api/LogDirFailureTest.scala [hunk 2]
```diff
@@ -33,7 +31,7 @@ import org.junit.Assert.assertTrue
 /**
   * Test whether clients can producer and consume when there is log directory failure
   */
-class LogDirFailureTest extends IntegrationTestHarness {
+class LogDirFailureTest() extends IntegrationTestHarness {
```
Comment (ijuma) [inline core/src/test/scala/integration/kafka/api/LogDirFailureTest.scala]: What's the reason for this change? `()` in a class without parameters is generally discouraged.

File: core/src/test/scala/integration/kafka/api/LogDirFailureTest.scala [hunk 3]
```diff
@@ -42,7 +40,6 @@ class LogDirFailureTest extends IntegrationTestHarness {
   this.logDirCount = 2
   this.producerConfig.setProperty(ProducerConfig.RETRIES_CONFIG, "0")
   this.producerConfig.setProperty(ProducerConfig.RETRY_BACKOFF_MS_CONFIG, "100")
-  this.serverConfig.setProperty(KafkaConfig.ReplicaHighWatermarkCheckpointIntervalMsProp, "100")
```
Comment (ijuma) [inline core/src/test/scala/integration/kafka/api/LogDirFailureTest.scala:45]: We are changing this so that it's easier to trigger the issue, correct? Do we want to remove it altogether or make it higher than before but still lower than the default?

File: core/src/main/scala/kafka/log/LogManager.scala [hunk 4]
```diff
@@ -216,7 +216,8 @@ class LogManager(logDirs: Array[File],
       maxProducerIdExpirationMs = maxPidExpirationMs,
       scheduler = scheduler,
       time = time,
-      brokerTopicStats = brokerTopicStats)
+      brokerTopicStats = brokerTopicStats,
+      logDirFailureChannel = logDirFailureChannel)
```
Comment (ijuma) [inline core/src/main/scala/kafka/log/LogManager.scala:220]: This should have been a compiler error in the first place. The default that we have `null` as a default argument in `Log.apply` doesn't seem right. What's the motivation for that?

**Review by ijuma — COMMENTED** [review]
Thanks for looking into it quickly and for the PR. The bug fix is clear. I left a few comments/questions on some of the other changes and the reason why we didn't catch this at compile time (as we should IMO).

File: core/src/main/scala/kafka/log/LogManager.scala (hunk 4 above)
Comment (lindong28) [reply core/src/main/scala/kafka/log/LogManager.scala:220]: @ijuma This is because there are only 2 invocation of `Log.apply` in source package whereas there are 61 invocation of `Log.apply` in test package. I thought it is simply to just use `null` as default value instead of editing 61 lines in the test package in order to reduce the code change in KIP-112.

Would you like me to remove this default value and change that 61 usage in the test package?

File: core/src/test/scala/integration/kafka/api/LogDirFailureTest.scala (hunk 2 above)
Comment (lindong28) [reply core/src/test/scala/integration/kafka/api/LogDirFailureTest.scala]: My bad. This change is a result of my making this class Parameterized and reverting that later. I will remove this `()`.

File: core/src/test/scala/integration/kafka/api/LogDirFailureTest.scala (hunk 3 above)
Comment (lindong28) [reply core/src/test/scala/integration/kafka/api/LogDirFailureTest.scala:45]: Previously this interval is reduced to 100 ms so that it can be triggered faster in the test. But this is no longer necessary now that the test explicitly calls either `replicaManager.checkpointHighWatermarks()` or `roll()` instead of relying on the periodic checkpoint operation.

Currently the default value of `ReplicaHighWatermarkCheckpointIntervalMs` is 5 sec. I think it is better to set it to 60 seconds so that the checkpoint is only invoked when the test explicitly invokes it. Does this sound OK?

File: core/src/main/scala/kafka/log/LogManager.scala (hunk 1 above)
Comment (lindong28) [reply core/src/main/scala/kafka/log/LogManager.scala]: Sure. I have replaced these multiple lines of logging with the logging in `LogDirFailureChannel.maybeAddLogFailureEvent()`. Can you check if it addresses the problem?

File: core/src/test/scala/integration/kafka/api/LogDirFailureTest.scala (hunk 3 above)
Comment (lindong28) [inline core/src/test/scala/integration/kafka/api/LogDirFailureTest.scala]: Fixed now.

File: core/src/test/scala/integration/kafka/api/LogDirFailureTest.scala (hunk 2 above)
Comment (lindong28) [reply core/src/test/scala/integration/kafka/api/LogDirFailureTest.scala]: It is fixed now. Thanks.

File: core/src/main/scala/kafka/log/LogManager.scala (hunk 4 above)
Comment (lindong28) [reply core/src/main/scala/kafka/log/LogManager.scala:220]: And thanks much for the comment @ijuma

Comment (ijuma) [reply core/src/main/scala/kafka/log/LogManager.scala:220]: I see. In general, we should do our best to avoid making production code more brittle due to tests. I would personally remove all the default arguments from `Log.apply`. By far most of the usages are from `LogTest`, so we can simply add a method within that class to make the migration easier.

Comment (ijuma) [reply core/src/main/scala/kafka/log/LogManager.scala:220]: Also, to be clear, I understand that the JBOD PR sort of just followed the existing pattern in that class.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6447/
Test PASSed (JDK 7 and Scala 2.11).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6432/
Test PASSed (JDK 8 and Scala 2.12).

File: core/src/main/scala/kafka/log/LogManager.scala (hunk 4 above)
Comment (lindong28) [reply core/src/main/scala/kafka/log/LogManager.scala:220]: @ijuma Sure. I have updated the patch to remove all default arguments in `Log.apply`. Can you take a look? Thanks!

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6457/
Test PASSed (JDK 7 and Scala 2.11).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6442/
Test PASSed (JDK 8 and Scala 2.12).

File: core/src/test/scala/unit/kafka/log/LogTest.scala [hunk 5]
```diff
@@ -2344,16 +2249,18 @@ class LogTest {
   @Test
   def shouldNotDeleteSegmentsWhenPolicyDoesNotIncludeDelete() {
     def createRecords = TestUtils.singletonRecords("test".getBytes, key = "test".getBytes(), timestamp = 10L)
-    val log = createLog(createRecords.sizeInBytes,
-      retentionMs = 10000,
-      cleanupPolicy = "compact")
+    val logProps = new Properties()
+    logProps.put(LogConfig.SegmentBytesProp, createRecords.sizeInBytes * 5: Integer)
+    logProps.put(LogConfig.RetentionMsProp, 10000: Integer)
+    logProps.put(LogConfig.CleanupPolicyProp, "compact")
```
Comment (ijuma) [inline core/src/test/scala/unit/kafka/log/LogTest.scala]: It seems like we could introduce a `logProps` method to keep things concise. Something along the lines of:

```scala
def logProps(messageSizeInBytes: Int, retentionMs: Int = -1, retentionBytes: Int = -1, ...)
    val logProps = new Properties()
    logProps.put(LogConfig.SegmentBytesProp, messageSizeInBytes * messagesPerSegment: Integer)
    logProps.put(LogConfig.RetentionMsProp, retentionMs: Integer)
    logProps.put(LogConfig.RetentionBytesProp, retentionBytes: Integer)
    logProps.put(LogConfig.CleanupPolicyProp, cleanupPolicy)
    logProps.put(LogConfig.MessageTimestampDifferenceMaxMsProp, Long.MaxValue.toString)
    logProps
}
```

What do you think?

File: core/src/test/scala/unit/kafka/log/LogCleanerManagerTest.scala [hunk 6]
```diff
@@ -236,13 +236,17 @@ class LogCleanerManagerTest extends JUnitSuite with Logging {
       recoveryPoint = 0L,
       scheduler = time.scheduler,
       time = time,
-      brokerTopicStats = new BrokerTopicStats)
+      brokerTopicStats = new BrokerTopicStats,
+      maxProducerIdExpirationMs = 60 * 60 * 1000,
+      producerIdExpirationCheckIntervalMs = 10 * 60 * 1000,
+      logDirFailureChannel = null)
```
Comment (ijuma) [inline core/src/test/scala/unit/kafka/log/LogCleanerManagerTest.scala]: Should we not pass an instance of `LogDirFailureChannel` instead of `null`? It seems a bit odd to pass an invalid value. If we added some validation to the apply method this test would start failing. Same for every other place we currently pass `null`.

File: core/src/test/scala/integration/kafka/api/LogDirFailureTest.scala [hunk 7]
```diff
@@ -123,3 +143,7 @@ class LogDirFailureTest extends IntegrationTestHarness {
   }
 
 }
+
+sealed trait LogDirFailureType
+case object Roll extends LogDirFailureType
+case object Checkpoint extends LogDirFailureType
```
Comment (ijuma) [inline core/src/test/scala/integration/kafka/api/LogDirFailureTest.scala]: This should be in a companion object since it seems specific to this test.

File: core/src/main/scala/kafka/server/LogDirFailureChannel.scala [hunk 8]
```diff
@@ -38,7 +41,8 @@ class LogDirFailureChannel(logDirNum: Int) {
    * If the given logDir is not already offline, add it to the
    * set of offline log dirs and enqueue it to the logDirFailureEvent queue
    */
-  def maybeAddLogFailureEvent(logDir: String): Unit = {
+  def maybeAddLogFailureEvent(logDir: String, msg: => String, t: IOException): Unit = {
```
Comment (ijuma) [inline core/src/main/scala/kafka/server/LogDirFailureChannel.scala]: Since the exception is optional, it should be be `Option[IOException]`. We should name it `exception` instead of `t`. Alternatively, we could have an overloaded method that doesn't take an exception.

File: core/src/main/scala/kafka/server/LogDirFailureChannel.scala [hunk 9]
```diff
@@ -29,7 +32,7 @@ import java.util.concurrent.{ArrayBlockingQueue, ConcurrentHashMap}
  * can take the name of the new offline log directory out of the LogDirFailureChannel and handles the log failure properly.
  *
  */
-class LogDirFailureChannel(logDirNum: Int) {
+class LogDirFailureChannel(logDirNum: Int) extends Logging {
 
   private val offlineLogDirs = new ConcurrentHashMap[String, String]
   private val logDirFailureEvent = new ArrayBlockingQueue[String](logDirNum)
```
Comment (ijuma) [inline core/src/main/scala/kafka/server/LogDirFailureChannel.scala]: Not related to this PR, but it's quite confusing that a queue of log dirs is called `logDirFailureEvent`. It sounds like it contains events instead of log dirs.

File: core/src/main/scala/kafka/server/LogDirFailureChannel.scala [hunk 10]
```diff
@@ -29,7 +32,7 @@ import java.util.concurrent.{ArrayBlockingQueue, ConcurrentHashMap}
  * can take the name of the new offline log directory out of the LogDirFailureChannel and handles the log failure properly.
```
Comment (ijuma) [inline core/src/main/scala/kafka/server/LogDirFailureChannel.scala]: Not related to this PR, but it would be good to clarify that log dirs are added only once and they are deemed offline until a restart.

File: core/src/main/scala/kafka/server/LogDirFailureChannel.scala [hunk 11]
```diff
@@ -38,7 +41,8 @@ class LogDirFailureChannel(logDirNum: Int) {
    * If the given logDir is not already offline, add it to the
    * set of offline log dirs and enqueue it to the logDirFailureEvent queue
```
Comment (ijuma) [inline core/src/main/scala/kafka/server/LogDirFailureChannel.scala:44]: Note related to this PR: also worth adding a sentence stating that once a directory is offline, it stays that way until a restart.

File: core/src/main/scala/kafka/server/LogDirFailureChannel.scala (hunk 8 above)
Comment (ijuma) [inline core/src/main/scala/kafka/server/LogDirFailureChannel.scala]: Not related to this PR, but it would probably be better if `logDir` was a `Path` (or `File`, but the former is the recommended one since Java 7). Generally, it's better to use types other than `String`. Same for `takeNextLogFailureEvent`.

File: core/src/main/scala/kafka/log/LogManager.scala [hunk 12]
```diff
@@ -307,7 +308,7 @@ class LogManager(logDirs: Array[File],
             error(s"Error while deleting the clean shutdown file $cleanShutdownFile", e)
         }
       }
-      offlineDirs.foreach(logDirFailureChannel.maybeAddLogFailureEvent)
+      offlineDirs.foreach(dir => logDirFailureChannel.maybeAddLogFailureEvent(dir, "Error while deleting the clean shutdown file", null))
```
Comment (ijuma) [inline core/src/main/scala/kafka/log/LogManager.scala]: We should include `dir` in the error message.

**Review by ijuma — COMMENTED** [review]
Thanks for the update Dong. Looks good overall, left a few additional comments. For the "not related to this PR" comments, they could be addressed separately if you prefer.

File: core/src/test/scala/unit/kafka/log/LogTest.scala (hunk 5 above)
Comment (lindong28) [reply core/src/test/scala/unit/kafka/log/LogTest.scala]: Good idea. It is just a lot of edit.. Anyway, I have fixed this as suggested.

File: core/src/test/scala/unit/kafka/log/LogCleanerManagerTest.scala (hunk 6 above)
Comment (lindong28) [reply core/src/test/scala/unit/kafka/log/LogCleanerManagerTest.scala]: In this particular case I think it is OK to pass null simply because it is not expected to be used.. if it is then it will throw NullPointerException and we will discover this when test fails.

Anyway, I have updated the patch to replace null with `new LogDirFailureChannel(10)`. It also works for me.

File: core/src/test/scala/integration/kafka/api/LogDirFailureTest.scala (hunk 7 above)
Comment (lindong28) [reply core/src/test/scala/integration/kafka/api/LogDirFailureTest.scala]: Good point. Thanks. I have fixed this.

File: core/src/main/scala/kafka/server/LogDirFailureChannel.scala (hunk 8 above)
Comment (lindong28) [reply core/src/main/scala/kafka/server/LogDirFailureChannel.scala]: I have updated the code to make this exception a required argument.

File: core/src/main/scala/kafka/log/LogManager.scala (hunk 12 above)
Comment (lindong28) [reply core/src/main/scala/kafka/log/LogManager.scala]: Sure. Fixed now.

File: core/src/main/scala/kafka/server/LogDirFailureChannel.scala (hunk 10 above)
Comment (lindong28) [reply core/src/main/scala/kafka/server/LogDirFailureChannel.scala]: Sure. I updated the comment to the following:

```
 * LogDirFailureChannel allows an external thread to block waiting for new offline log dir.
 *
 * LogDirFailureChannel should be a singleton object which can be accessed by any class that does disk-IO operation.
 * If IOException is encountered while accessing a log directory, the corresponding class can add the the log directory name
 * to the LogDirFailureChannel using maybeAddLogFailureEvent(). Each log directory can be added only once. After a log directory
 * is added for the first time, a thread which is blocked waiting for new offline log directories
 * can take the name of the new offline log directory out of the LogDirFailureChannel and handles the log failure properly. An
 * offline log directory will stay offline until a broker restarts.
```

File: core/src/main/scala/kafka/server/LogDirFailureChannel.scala (hunk 9 above)
Comment (lindong28) [reply core/src/main/scala/kafka/server/LogDirFailureChannel.scala]: It was named as event because we may expand it later to include more information than the log directory name. I have updated the code to rename methods as suggested.

File: core/src/main/scala/kafka/server/LogDirFailureChannel.scala (hunk 11 above)
Comment (lindong28) [reply core/src/main/scala/kafka/server/LogDirFailureChannel.scala:44]: Sure. It is added now.

**Comment by lindong28** [comment]
@ijuma Thanks much for the suggestion. All comments have been addresses as suggested. Could you take another look?

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6490/
Test PASSed (JDK 7 and Scala 2.11).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6475/
Test PASSed (JDK 8 and Scala 2.12).

File: core/src/main/scala/kafka/server/LogDirFailureChannel.scala [hunk 13]
```diff
@@ -27,33 +27,35 @@ import kafka.utils.Logging
  * LogDirFailureChannel allows an external thread to block waiting for new offline log dir.
  *
  * LogDirFailureChannel should be a singleton object which can be accessed by any class that does disk-IO operation.
- * If IOException is encountered while accessing a log directory, the corresponding class can insert the the log directory name
- * to the LogDirFailureChannel using maybeAddLogFailureEvent(). Then a thread which is blocked waiting for new offline log directories
- * can take the name of the new offline log directory out of the LogDirFailureChannel and handles the log failure properly.
+ * If IOException is encountered while accessing a log directory, the corresponding class can add the the log directory name
+ * to the LogDirFailureChannel using maybeAddLogFailureEvent(). Each log directory can be added only once. After a log directory
```
Comment (ijuma) [inline core/src/main/scala/kafka/server/LogDirFailureChannel.scala:31]: I'll update this to use the new method name before merging: maybeAddOfflineLogDir. Also tweaked the wording a bit.

File: core/src/main/scala/kafka/log/LogManager.scala [hunk 14]
```diff
@@ -214,9 +213,11 @@ class LogManager(logDirs: Array[File],
       logStartOffset = logStartOffset,
       recoveryPoint = logRecoveryPoint,
       maxProducerIdExpirationMs = maxPidExpirationMs,
+      producerIdExpirationCheckIntervalMs = 10 * 60 * 1000,
```
Comment (ijuma) [inline core/src/main/scala/kafka/log/LogManager.scala:216]: I introduced a constant for this and used it everywhere else.

File: core/src/test/scala/integration/kafka/api/LogDirFailureTest.scala [hunk 15]
```diff
@@ -123,3 +144,10 @@ class LogDirFailureTest extends IntegrationTestHarness {
   }
 
 }
+
+object LogFailureTest {
```
Comment (ijuma) [inline core/src/test/scala/integration/kafka/api/LogDirFailureTest.scala:148]: Renamed to LogDirFailureTest.

File: core/src/test/scala/unit/kafka/log/LogTest.scala [hunk 16]
```diff
@@ -3019,26 +2812,52 @@ class LogTest {
...
+                      segmentBytes: Int = 1 * 1024 * 1024 * 1024,
+                      retentionMs: Long = 24 * 7 * 60 * 60 * 1000L,
+                      retentionBytes: Long = -1L,
+                      segmentJitterMs: Long = 0L,
+                      cleanupPolicy: String = "delete",
+                      maxMessageBytes: Int = 1000000 + MessageSet.LogOverhead,
+                      indexIntervalBytes: Int = 4096,
+                      segmentIndexBytes: Int = 10 * 1024 * 1024,
+                      messageFormatVersion: String = ApiVersion.latestVersion.toString,
+                      fileDeleteDelayMs: Long = 60000L): LogConfig = {
```
Comment (ijuma) [inline core/src/test/scala/unit/kafka/log/LogTest.scala:2825]: Replaced these hardcoded values by the constants in `Defaults`.

File: core/src/test/scala/unit/kafka/log/LogTest.scala [hunk 17]
```diff
@@ -1605,15 +1504,14 @@ class LogTest {
     }
 
     // reopen the log
-    log = Log(logDir, config, logStartOffset = 0L, recoveryPoint = 200L, scheduler = time.scheduler,
-      brokerTopicStats = brokerTopicStats, time = time)
+    log = createLog(logDir, logConfig)
```
Comment (ijuma) [inline core/src/test/scala/unit/kafka/log/LogTest.scala:1507]: Seems like `recoveryPoint = 200L` was lost here, I added it back.

File: core/src/test/scala/unit/kafka/log/LogTest.scala [hunk 18]
```diff
@@ -1460,14 +1384,12 @@ class LogTest {
       assertEquals("Should have same number of time index entries as before.", numTimeIndexEntries, log.activeSegment.timeIndex.entries)
     }
 
-    log = Log(logDir, config, logStartOffset = 0L, recoveryPoint = lastOffset, scheduler = time.scheduler,
-      brokerTopicStats = brokerTopicStats, time = time)
+    log = createLog(logDir, logConfig)
```
Comment (ijuma) [inline core/src/test/scala/unit/kafka/log/LogTest.scala:1387]: Looks like `recoveryPoint = lastOffset` was lost here. I added it back.

File: core/src/test/scala/unit/kafka/log/LogTest.scala [hunk 19]
```diff
@@ -1046,22 +1005,18 @@ class LogTest {
...
   }
 
   /**
    * This test case appends a bunch of messages and checks that we can read them all back using sequential offsets.
    */
   @Test
   def testAppendAndReadWithSequentialOffsets() {
-    val logProps = new Properties()
-    logProps.put(LogConfig.SegmentBytesProp, 71: java.lang.Integer)
     // We use need to use magic value 1 here because the test is message size sensitive.
```
Comment (ijuma) [inline core/src/test/scala/unit/kafka/log/LogTest.scala:1017]: I removed this comment since we were not and are not using message format V1 (existing issue).

File: core/src/test/scala/unit/kafka/log/LogTest.scala [hunk 20]
```diff
@@ -1331,12 +1268,9 @@ class LogTest {
     val messageSet = MemoryRecords.withRecords(CompressionType.NONE, new SimpleRecord("You".getBytes), new SimpleRecord("bethe".getBytes))
     // append messages to log
     val configSegmentSize = messageSet.sizeInBytes - 1
-    val logProps = new Properties()
-    logProps.put(LogConfig.SegmentBytesProp, configSegmentSize: java.lang.Integer)
     // We use need to use magic value 1 here because the test is message size sensitive.
```
Comment (ijuma) [inline core/src/test/scala/unit/kafka/log/LogTest.scala:1271]: I removed this comment since it's wrong (existing issue).

**Review by ijuma — APPROVED** [review]
Thanks for the PR and for the clean-ups, LGTM. I found some minor issues mentioned below that I fixed before merging. I also updated the commit message to include more detail. I hope that's OK.

Merging to trunk.

**Comment by lindong28** [comment]
@ijuma Thanks a lot for your suggestion and for helping improve this patch!
